Lucene 10577

msokolov opened the pull request:
This is SCRATCH - not to be committed. It has numerous problems, but was useful for testing and I share it as a first broken impl that can be improved.  Things TBD:

1. work out a better way to figure out scaling (maybe let customer pass in 8-bit values, perhaps *as* floats).
2. do the vector math directly on the mmaped bytes using ByteVector
3. fix the tests so they can handle quantized data better

romseygeek committed:
Add next minor version 9.2.0

Commit:
LPVQ00

Commit:
clip when scaling; comment out some tests

rmuir commented on lucene/core/src/java/org/apache/lucene/codecs/lucene92/ExpandingRandomAccessVectorValues.java:
```diff
@@ -0,0 +1,57 @@
+        BytesRef binaryValue = delegateValues.binaryValue(targetOrd);
+        byte[] bytes = binaryValue.bytes;
+        for (int i = 0, j = binaryValue.offset; i < value.length; i++, j++) {
+          value[i] = bytes[j] * scale;
```
Seems to me that moving dotProduct etc out of `org.apache.lucene.util` could help. It could be in the codec.

at a glance, i would modify dotproduct vectors patch and try something like:
```
FloatVector floats = ByteVector.fromArray(bytes).reinterpretAsFloats();
floats = floats.mul(scale);
... remainder of existing algorithm from patch ...
```

I have no idea how this would perform off the top of my head, but we can try it.

rmuir replied:
and i think we don't want reinterpret, but this one: https://docs.oracle.com/en/java/javase/16/docs/api/jdk.incubator.vector/jdk/incubator/vector/ByteVector.html#viewAsFloatingLanes()

rmuir replied:
the javadoc illustrates the challenge: "This method always throws UnsupportedOperationException, because there is no floating point type of the same size as byte. The return type of this method is arbitrarily designated as Vector<?>. Future versions of this API may change the return type if additional floating point types become available."

rmuir replied:
I'm not really seeing the path to vectorization here though. At least not based on what i see here. Like i said, I really do think it should block the issue because if we add support to the index format it is difficult/impossible to get rid of.

rmuir replied:
I think i'd be able to help more, if the current PR was actually doing said byte-based formula and not converting to floats.

The way the code is structured now (doing floating point based math), I can't really help, it won't be efficient.

msokolov replied:
Yeah - I hear you. I will try to work out more concretely what that would look like. My initial preoccupation has been with (1) a reasonable API and (2) ensuring that 8 bits is enough to get good recall.

msokolov replied:
I did some git booboo and messed this up -- I'll close and open a new PR since it's a substantially different approach anyway

mocobeta committed:
LUCENE-10531: Add @RequiresGUI test group for GUI tests (#893)

Co-authored-by: Dawid Weiss <dawid.weiss@carrotsearch.com>

mocobeta committed:
LUCENE-10531: small follow-up for b911d1d47

jpountz committed:
LUCENE-9409: Reenable TestAllFilesDetectTruncation. (#896)

- Removed dependency on LineFileDocs to improve reproducibility.
 - Relaxed the expected exception type: any exception is ok.
 - Ignore rare cases when a file still appears to have a well-formed footer
   after truncation.

jpountz committed:
LUCENE-10574: Prevent pathological merging. (#900)

This updates TieredMergePolicy and Log(Doc|Size)MergePolicy to only ever
consider merges where the resulting segment would be at least 50% bigger than
the biggest input segment. While a merge that only grows the biggest segment by
50% is still quite inefficient, this constraint is good enough to prevent
pathological O(N^2) merging.

jpountz committed:
LUCENE-10574: Fix test failure.

If a LogByteSizeMergePolicy is used, then it might decide to not merge the two
one-document segments if their on-disk sizes are too different. Using a
LogDocMergePolicy addresses the issue as both segments are always considered
the same size.

jpountz committed:
LUCENE-10574: Fix test failure.

LogDocMergePolicy would previously always force-merge an index that has 10
segments of size 1 to 10, due to the min doc count. This is not the case
anymore, but the test was assuming that such an index would get merged, so I
fixed the test's expectations.

Also changed the merge policy to keep working when RAM buffers are flushed in
such a way that segments do not appear in decreasing size order using the same
logic as LogMergePolicy.

jpountz committed:
LUCENE-10574: More test failures.

 - MergeOnFlushMergePolicy doesn't try to avoid O(n^2) merges, so I'm disabling
   the test on it for now.
 - TestUpgradeIndexMergePolicy would sometimes wrap with a non-standard merge
   policy like the alcoholic merge policy, I forced it to wrap a
   TieredMergePolicy.

jpountz committed:
Fix precommit.

Commit:
make HnswGraph utils generic; handle BytesRef vectors

jpountz committed:
LUCENE-10574: Keep allowing unbalanced merges if they would reclaim lots of deletes. (#905)

`TestTieredMergePolicy` caught this special case: if a segment has lots of
deletes, we should still allow unbalanced merges.

rmuir committed:
LUCENE-10579: fix smoketester backwards-check to not parse stdout (#903)

This is very noisy, can contain gradle status updates, various other tests.verbose prints from other threads, you name it.

It causes the check to be flaky, and randomly "miss" seeing a test that executed.

Instead, let's look at the zip files. We can still preserve the essence of what the test wants to do, but without any flakiness.

Commit:
refactor findWorstNonDiverse; fix tests

Commit:
convert query to bytes if needed in HnswGraphSearcher

Commit:
nocommits

Commit:
no CFS

Commit:
fixing unused, etc

Commit:
spotless
